상품 CRUD와 사용자 정보를 세션으로 연결, 글쓴이 프로필 정보 연동

ahyeonkong opened the pull request:
## #️⃣ Issue Number

#70 

## 📝 작업 설명

- 상품 조회 페이지에서 사용자 프로필 정보가 같이 뜨도록 구현합니다.
상품을 등록한 사람의 프로필 정보가 나타납니다.
로그인 세션을 반영합니다.

- HttpSession session을 통해 세션 값을 받아옵니다. ->Spring Security의 getSession()으로 변경했습니다.
- 프로필 조회라는 이름에 맞게 UserResponse를 UserProfileResponse로 변경합니다.
- 프로필 정보의 경우, 사용자의 id, nickname, profileUrl만 뜨도록 구현합니다. 주소는 사용자의 주소를 인증할 방법이 없어서 구현하지 않는 걸로 알고 있습니다!
- 세션을 통해 사용자 정보를 얻으므로 request에서 불필요한 user 필드를 제거합니다. 기존에는 user의 id를 직접 입력받는 코드로 구현이 되어 있었는데, 세션을 통해 받아오기 때문에 해당 코드는 삭제했습니다.
---
(여기부터 수정입니다! 😁)
- Spring Security의 getSession을 통해 세션을 가져옵니다. 기존의 SessionUtil 클래스는 내용이 중복되므로 삭제합니다.
- 로그인 여부에 따라 확인 가능한 페이지를 제한했습니다.
로그인 필수 X: 상품 목록 조회(상품 등록과 경로가 겹쳐서 "/list" 추가)
로그인 필수 O: 상품 등록, 상품 수정(글쓴이만 가능), 상품 삭제(글쓴이만 가능), 상품 상세 조회, 상품 끌어올리기(글쓴이만 가능), 상품 좋아요(좋아요 추가/취소에 대한 응답 SuccessCode 추가)


<details>
<summary>세션 방식에 대한 설명입니다 🤩</summary>
<div>

### 세션을 이용한 시큐리티 동작 방식
 1. 사용자가 요청을 보내면 Spring Security 필터 체인이 동작
 2. SecurityContextPersistenceFilter가 세션에서 인증 정보를 찾아 SecurityContextHolder에 설정
 3. 컨트롤러에서 세션 정보 확인

### 세션을 이용한 시큐리티 설명
**이전 방식(HttpSession httpSession 매개변수 주입)**
- 세션 관리를 직접 구현
- 보안 관련 기능 직접 구현 필요
- 단순하지만 기능 제한적 
- session.getAttribute() 직접 호출
- 타입 캐스팅 필요
- 매번 null 체크 필요
- CSRF 보호 없음
- 세션 고정 공격 취약
- 권한 관리 직접 구현
- **경로 변수({id}) 처리가 복잡** -> 이 문제 때문에 JWT 토큰 발급으로 변경하려 했지만, Spring Security를 사용하기로 결정.

**개선 방식(Spring Security 활용)**
- 세션 관리 자동화
- 보안 기능 기본 제공
- 권한 관리 용이
- @AuthenticationPrincipal 사용 가능
- SecurityContextHolder 통해 어디서든 접근 가능
- 타입 안전성 보장
- CSRF 보호 기본 제공
- 세션 고정 방지
- 권한 관리 프레임워크 제공

### Spring Security 방식을 활용했을 때의 이점
1. 로그인 API 호출
2. JSESSIONID 자동 관리
3. 이후 요청에 자동으로 세션 포함

</div>
</details>

## 📸스크린샷 (선택)   
상품 상세 조회 API 응답값입니다.
![image](https://github.com/user-attachments/assets/24ae07b8-1e6b-49a0-8e97-c51dabbefa16)


## 💬 리뷰 요구사항

- profileUrl의 경우, 위의 사진에 나타나듯이 파일 명을 보내줄지 아니면 프리사인 url로 접근할지 정하면 될 것 같습니다.
- 현재 상품 관련 post 요청의 응답 값과 get 요청의 응답 값이 같은 것 같습니다! 상세 정보의 경우, get 요청에서만 응답 값을 넘겨줘도 되지 않을까 하는 생각입니다!!

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

ahyeonkong committed:
fix: 인증/미인증 경로 수정

ahyeonkong committed:
refactor: 상품 등록 시 유저 정보 연결

ahyeonkong committed:
refactor: 유저 프로필 관련 이름으로 변경

ahyeonkong committed:
refactor: UserResponse -> UserProfileResponse 이름 변경

ahyeonkong committed:
refactor: ApiResponse를 통한 성공/에러 응답 통일

ahyeonkong committed:
fix: 인증 경로 수정

ahyeonkong committed:
refactor: SecurityContext를 이용한 인증 정보 활용

- User 엔티티의 getAuthorities()를 통한 사용자 권한 확인
- 세션에 SecurityContext와 사용자 정보 저장
- 통합 조회 API를 @GetMapping(/list)로 개선
- 매개변수 httpSession -> getSession() 방식으로 변경
- 상품 수정, 삭제의 경우 글쓴이만 가능하도록 설정

ahyeonkong committed:
fix: 불필요한 에러 처리 삭제

ahyeonkong committed:
feat: 사용자 정보를 담는 UserDetails 메서드 구현

ahyeonkong committed:
refactor: ID로 상품 조회 후 존재하지 않는 상품 예외 처리

ahyeonkong committed:
fix: 상품 끌어올리기 경로 수정

ahyeonkong committed:
refactor: 끌어올리기 세션 연결 및 글쓴이만 가능하도록 제한

ahyeonkong committed:
fix: 소유자 검증 중복 로직 제거

ahyeonkong committed:
feat: 좋아요 추가/취소 응답 생성

ahyeonkong committed:
refactor: 좋아요 기능에 세션 연결 방법 개선 및 좋아요 추가/취소 응답 생성

JunBe commented on src/main/java/com/market/saessag/global/config/PathConst.java:
```diff
@@ -6,18 +6,20 @@ public class PathConst {
-            "/api/products/{id}/update",       // 상품 수정
-            "/api/products/{id}/delete",       // 상품 삭제
+            "/api/products",                    // 상품 등록
+            "/api/products/**",                 // 상품 수정, 삭제, 상세 조회
```
`/api/products/**` 경로가 인증이 필요하면 `/api/products` 가 포함되어 있는 모든 경로에서 인증이 필요할 것 같습니다!
`/api/products/{id}` 와 같이 특정 경로만 인증하도록 변경하는 건 어떨까요??

ahyeonkong replied:
처음에 `/api/products/{id}`와 같이 경로를 설정했었는데 에러가 발생했습니다!
Spring Security의 URL 패턴 매칭 방식 때문에 `/api/products/**`와 같이 지정해야 한다고 합니다.

추가로 `/api/products`가 포함되어 있는 모든 경로에서 인증이 필요할 것 같다고 말씀 주셨는데, 상품 목록 조회의 경우 로그인을 하지 않아도 확인 가능하도록 구현했기 때문에 일부 경로에서만 인증이 필요하도록 만들 수 있습니다!

JunBe replied:
말을 오해하도록 전달한 것 같습니다 ㅠ.ㅠ 

`/api/products/**` 로 설정 하면 아래의 
```java
"/api/products/bump/**",            // 상품 끌어올리기
"/api/products/{id}/like",          // 상품 좋아요
"/api/products/changeStatus"       // 상품 상태 값 변경
```
경로들도 자동으로 인증이 필요하도록 설정 되지 않나 라는 궁금증이 있었습니다!

ahyeonkong replied:
아하 ㅋㅋㅋ 맞습니다!!
위처럼 하면 더 간단하게 나타낼 수 있을 것 같네요! 수정하겠습니다

JunBe commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -50,23 +60,67 @@ public ApiResponse<ProductResponse> getProductDetail(@PathVariable Long id,
+            @RequestBody ProductRequest productRequest,
+            HttpServletRequest request) {
+
+        HttpSession session = request.getSession();
```
유저 세션을 받아오고 로그인 한 유저와 상품의 소유자가 동일한지 판단하는 로직이 상품 수정, 상품 삭제에 중복되어 있습니다.
이 부분을 따로 메서드로 분리해서 사용하면 중복도 줄이고 나중에 또 다른 기능에서 필요하다면 편리하게 사용할 수 있을 것 같습니다!

ahyeonkong replied:
넵! 저도 이 부분은 동의합니다.
시간이 된다면 나중에 한번에 리팩토링 하는 게 좋을 것 같습니다!

JunBe commented on src/main/java/com/market/saessag/domain/user/controller/SignInController.java:
```diff
@@ -24,15 +30,37 @@ public class SignInController {
 
     @PostMapping("/sign-in")
     public ResponseEntity<ApiResponse<SignInResponse>> signIn(
```
ProductController와 동일하에 반환 타입 ApiResponse<?>로 통일하면 좋을 것 같습니다!

JinTaekLim replied:
반환타입은 가능하면 현재처럼 직접 명시하는 게 좋을 것같습니다!

ahyeonkong replied:
저도 직접 명시가 좋다고 생각합니다!
몇몇 코드는 아직 반환 타입 ApiResponse<?>로 되어있던데 수정하면 좋을 것 같습니다!

JunBe replied:
그럼 상품 쪽을 수정해 보겠습니다!

JinTaekLim commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -50,23 +60,67 @@ public ApiResponse<ProductResponse> getProductDetail(@PathVariable Long id,
+    public ApiResponse<ProductResponse> createProduct(@RequestBody ProductRequest productRequest,
+                                                      HttpServletRequest request) {
+        HttpSession session = request.getSession();
+        SignInResponse userSession = (SignInResponse) session.getAttribute("userProfile");
```
해당 로직은 service 단으로 이동하는 건 어떨까요?

ahyeonkong replied:
좋습니다!

JinTaekLim commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -50,23 +60,67 @@ public ApiResponse<ProductResponse> getProductDetail(@PathVariable Long id,
+    public ApiResponse<ProductResponse> updateProduct(
+            @PathVariable Long id,
+            @RequestBody ProductRequest productRequest,
+            HttpServletRequest request) {
```
RequestBody의 ProductRequest 안에 모든 매개변수를 넣어서 처리하지 않고 분리한 이유가 있을까요?

ahyeonkong replied:
`@PutMapping("/products/{id}")`에서 id를 RequestBody가 아닌 @PathVariable로 분리하여 사용하는 것은 RESTful API 설계 원칙에 따라 리소스를 URI로 명확하게 식별하기 위함입니다!

ahyeonkong committed:
refactor: 상품 관련 인증 경로 간소화

ahyeonkong committed:
refactor: 상품 생성 api의 세션 처리 로직을 서비스 계층으로 이동

- 컨트롤러는 요청/응답 처리만 책임

ahyeonkong committed:
refactor: 상품 수정 api의 인증 및 권한 검증 로직을 서비스 계층으로 이동

- 컨트롤러의 역할 단순화

ahyeonkong committed:
refactor: 상품 삭제 api의 인증 및 권한 검증 로직을 서비스 계층으로 이동

- 컨트롤러 역할 단순화

ahyeonkong committed:
fix: 상품 상태 변경 api 요청/응답 객체 직렬화 오류 수정

- 필요한 어노테이션 추가(@NoArgsConstructor, @AllArgsConstructor, @Getter)
- 필드명 변경 (productId -> id)

ahyeonkong committed:
refactor: 상품 상태 값 변경 api의 인증 및 권한 검증 로직 추가

- 상품 상태 값 변경 시 세션을 통한 사용자 인증 로직 추가
- 상품 소유자 검증 로직 추가하여 본인 상품만 수정 가능하도록 개선
- 예외 처리 추가(로그인되지 않은 요청, 타인의 상품 수정 시도, 존재하지 않는 상품)

ahyeonkong commented:
어느정도 리팩토링이 완료되었기 때문에 합치겠습니다!
